src/lib/FreeC/Pass/EtaConversionPass.hs

```diff
@@ -169,3 +238,26 @@ arityOf (IR.Undefined _ _      ) = return 0
 arityOf (IR.ErrorExpr  _ _ _   ) = return 0
 arityOf (IR.IntLiteral _ _ _   ) = return 0
 arityOf (IR.Lambda _ _ _ _     ) = return 0
+
+-------------------------------------------------------------------------------
+-- Helper functions                                                                     --
+-------------------------------------------------------------------------------
+
+-- | 'splitReturnType n' uncurries the given function's type n times.
+--   It is assumed that the given type is of the form
+--   @τ₁ -> (… -> (τₙ -> τ))@, which is true for the return type of 
+--   a partially defined function with @n@ missing arguments.
+--   The function returns a tuple consisting of a list of the argument
+--   types @[τ₁, …, τₙ]@ and the return type @τ@. 
+splitReturnType :: Int -> IR.Type -> ([IR.Type], IR.Type)
+splitReturnType 0 t = ([], t)
+splitReturnType n t =
+  let (arg , r    ) = IR.splitFuncType t 1
+      (args, rType) = splitReturnType (n - 1) r
+  in  (arg ++ args, rType)
```

Comment:
I don't get how this is different from `IR.splitFuncType` (except that the arguments are flipped).

Also, the assumption that the type has the form `τ₁ -> (… -> (τₙ -> τ))` is not necessarily true in the presence of type synonyms. The [`TypeSignaturePass`](https://github.com/FreeProving/free-compiler/blob/4627f930212624096a61ad051b557c0690f4d838/src/lib/FreeC/Pass/TypeSignaturePass.hs#L187) deals with this issue by expanding type synonyms on demand. I think that the most reliable solution would be to move the `etaConversionPass` above the `typeSignaturePass` in the compiler [`pipeline`](https://github.com/FreeProving/free-compiler/blob/4627f930212624096a61ad051b557c0690f4d838/src/lib/FreeC/Pipeline.hs#L25). This would have the advantage, that you wouldn't have to annotate any types or modify environment entries. However, this change would be more involved and is not really in the scope of a first issue. Thus, I'd suggest that we just recycle the definition of `FreeC.Pass.TypeSignaturePass.splitFuncType` for now.

Reply:
It isn't; I simply misunderstood what `IR.splitFuncType` actually does. I've changed it so that `etaConvertFuncDecl` simply uses `IR.splitFuncType` now.

Reply:
Ah, okay. If you have a suggestion for how to improve the documentation of `IR.splitFuncType` (or any other function), don't hesitate to rephrase it : )